### DISCUSSION TITLE
Versioning, Forking, Updating Burritos

### DISCUSSION THREAD

**Discussion by jtauber** [bible-technology/scripture-burrito#246] [Archive]
# Versioning, Forking, Updating Burritos

Recall that a burrito is a snapshot of a resource or project (henceforth "resource" where "resource or project" is to be understood) at a particular point in time. Changes may be made to a resource and, when this happens, a new burrito generated. It is important to view this as a new burrito. In other words, burritos are not edited, nor do they have a version history of their own. Rather a burrito reflects a snapshot of a resource which may itself have a version history. Furthermore, one may import a burrito generated in one system into another system, make edits to the resource, and then generate a new burrito reflecting the "forked" resourced. This new burrito may subsequently be imported back into the original system as a new version of the original resource. 

This section gives some concrete use cases with details of the relevant information tracked in the burrito. In these use cases, I assume that each "system" (software used for creating / editing the resource and generating the burrito) is its own authority for generating identifers (although in practice this may be delegated).

## Use Case 1: A New Resource

System A creates a resource and exports a burrito. The burrito contains the minted "System A" identifier along with a revision number and timestamp.

```
identity: A foo v1
```

NOTE: If the same version of a resource is exported again with no changes, we would expect the revision number and timestamp in the identifier section to remain the same (as well as all the content) although the burrito's own timestamp itself may have changed.

## Use Case 2: Revision to a Resource

System A makes changes to the resource above and exports another burrito. The burrito contains the same "System A" identifer as before but the revision number and timestamp have changed.

```
identity: A foo v2
```

There is no further version history in the burrito. 

## Use Case 3: Importing into Second System

System B imports the burrito generated by System A above as a new resource. Without making any content changes, it then generates a burrito of its own. 

NOTE: after discussions, it was suggested by @FoolRunning that rather than distinguish whether a change has been made or not, System A's id is relegated to the upstream list anyway. This makes this use case identical to the following one.

```
identity: B bar v1
upstream: A foo v2
```

## Use Case 4: Forking a Resource

System B imports the burrito generated by System A above as a new resource, makes edits to it and then generates a new burrito. It should mint its own "System B" identifier, revision number, and timestamp (like in Use Case 3) but it should include the original System A identifier, revision number, and timestamp information in a NEW section indicating the upstream source of the resource before any modifications.

```
identity: B bar v2
upstream: A foo v2
```

## Use Case 5: Merge Back Resource Changes

System A imports the above burrito from System B to incorporate the changes made to the resource. It can make use of the "upstream source" information in the burrito to know which resource to merge the changes back in to. The System A identifier remains the same but a new revision number and timestamp are used.

If a burrito is then generated, the System B identifier, revision number, and timestamp get demoted to the "upstream source" section. 

```
identity: A foo v3
upstream: B bar v2
```

It is possible System A may want to import the burrito as a NEW resource in which case it would mint a new identifier and this would essentially just be use case 4 with A and B swapped.

```
identity: A buz v1
upstream: B bar v2
```

## Use Case 6: Combining Resources

System C takes independent resources from System A and B and combines them into a single resource. A subsequently exported burrito contains a System C identifier, revision number, and timestamp and the the "upstream source" section contains information both for the System A resource and System B resource (like in use case 4 but twice).

```
identity: C bit v1
upstream: A foo v3; B wug v1
```

NOTE: after discussion, we decided NOT to distinguish between upstream identifiers that represent the same resource from those that represent independent resources that have both been used to make the new resource.

**Comment by jonathanrobie** [comment]
I really like this - obviously, details need to be worked out, but this is semantically clean.

We need to think this through for chains of 3-4 steps.  A->B->C gets merged back into A and also gets merged back into B.

We also need to think about whether fork/merge can happen along the way, e.g.  A->(B1, B2)->C, where C is a merge process, merging back into A or B1.  That may be a YAGNI, we should figure out if it is a requirement or not, the complexity of supporting this well may not be needed.

**Comment by FoolRunning** [comment]
The only thing I would change is that I think a created burrito should always demote all other known IDs to the "upstream source" section. I don't think there is any particular use to keeping an ID elevated when there hasn't been any changes and it simplifies the spec since you don't need to keep track of "did I change anything".

**Reply by jtauber** [reply]
Yeah, I like this a lot. Basically when a system exports a burrito, only the identifier(s) for which it's an authority remain untouched and everything else should just be moved to the "upstream source" section.

**Reply by jtauber** [reply]
I've updated the write up to reflect this which basically means there's no difference between Use Cases 3 and 4.

**Comment by jag3773** [comment]
A list of every "place", e.g. idAuthority, should be included in a "history", or "Upstream", or "Last known" section.

**Comment by jag3773** [comment]
Possibly Use Case 6 should be a `derived` burrito scenario.

**Comment by jtauber** [comment]
## OPEN QUESTION

I did have one further question but I may have answered it myself on reflection.

I was initially wondered if the following would be true:
  - an authority can only be used once across the main identity and upstream list
  - the main identity is just the authority associated with the system producing the burrito

However, I can now think of multiple reasons why this might not be true:

1. I might derive a resource from two (or more) other resources all with identifiers minted by me. (e.g. `A foo v5` + `A bar v2` + `A baz v3`)
2. id authorities (like Github) might be shared between systems (e.g. `github jtauber/my-resource some-sha` and `github jag3773/another-resource another-sha`)

So what constraints (if any) DO exist on multiple entries with the same authority but different ids OR same authority+ids but different version number.

**Reply by jag3773** [reply]
I think that we want to disallow "same authority+ids but different version number" but everything else should be permissible.

We aren't using this to keep track of history, we are just using this to keep track of "locations" that this burrito has been. It is possible that the burrito has a location history which includes 2+ distinct *projects* with a given authority.

**Reply by jtauber** [reply]
Yes, I think in practice same authority+id but different version number doesn't make sense but we should decide whether to actually enforce it in the spec.

I think the two distinct projects with a given authority is an important use case because you might get merging of projects or resources being transferred from one project to another.
